src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/LegacySyntaxNodeExtensions.NodeStack.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+
+namespace Microsoft.AspNetCore.Razor.Language.Legacy;
+
+internal static partial class LegacySyntaxNodeExtensions
+{
+    private struct NodeStack : IDisposable
+    {
+        [ThreadStatic]
+        private static SyntaxNode[]? g_nodeStack;
+
+        private readonly SyntaxNode[] _stack;
+        private int _stackPtr;
+
+        public NodeStack(IEnumerable<SyntaxNode> nodes)
+        {
+            _stack = g_nodeStack ??= new SyntaxNode[64];
```

Comment:
Nothing about `NodeStack` implies to me that it would be thread-affinitized. It should at least be documented.

Reply:
Sure!

Reply:
I've gone ahead and switched this over to a pool.